### ISSUE TITLE
2.7x decode slowdown with Gated DeltaNet when input_embeddings differ from vocabulary vectors

### ISSUE THREAD

**Issue by josaum** [ml-explore/mlx-lm#932]
## Summary

When using `input_embeddings` with Qwen3.5-35B-A3B (Gated DeltaNet + Attention hybrid), decode throughput drops from ~93 tok/s to ~34 tok/s if the embeddings are not exact vocabulary vectors. Even a 0.01 perturbation to real vocabulary embeddings triggers the slowdown.

This matters for **soft prompt injection** (prompt tuning / prefix tuning), where learned continuous vectors are prepended to token embeddings before the forward pass.

## Reproduction

```python
import copy, time
import mlx.core as mx
import mlx_lm
from mlx_lm.cache_prompt import make_prompt_cache
from mlx_lm.sample_utils import make_sampler

model, tokenizer = mlx_lm.load("NexVeridian/Qwen3.5-35B-A3B-4bit")
embed_dim = getattr(model.args, "hidden_size", 2048)
sampler = make_sampler(temp=0.0)

messages = [{"role": "user", "content": "Explain what an embedding is."}]
prompt_str = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
tokens = mx.array(tokenizer.encode(prompt_str))
_base = model.language_model.model
token_embs = _base.embed_tokens(tokens)


def raw_decode(kv_cache, first_tok, max_tokens=64):
    y = mx.array([[first_tok]])
    toks = [first_tok]
    eos = tokenizer.eos_token_id
    for _ in range(max_tokens - 1):
        out = model(y, cache=kv_cache)
        y = sampler(out[:, -1:, :])
        mx.eval(y)
        t = y.item()
        if t == eos:
            break
        toks.append(t)
        y = y.reshape(1, 1)
    return toks


def make_cache_with_embeddings(embeddings):
    cache = make_prompt_cache(model)
    pad_ids = mx.full((1, len(embeddings)), 0, dtype=mx.int32)
    logits = model(pad_ids, cache=cache, input_embeddings=embeddings[None])
    mx.eval([c.state for c in cache])
    first = sampler(logits[:, -1:, :]).item()
    return cache, first


# --- Test A: exact vocab embeddings ---
pad_embs = _base.embed_tokens(mx.full((4,), tokenizer.pad_token_id or 0, dtype=mx.int32))
combined_vocab = mx.concatenate([pad_embs, token_embs], axis=0)
cache_a, first_a = make_cache_with_embeddings(combined_vocab)

# --- Test B: vocab embeddings + 0.01 noise ---
import numpy as np
rng = np.random.default_rng(42)
noise = mx.array(rng.standard_normal((4, embed_dim)).astype(np.float32)) * 0.01
combined_noisy = mx.concatenate([pad_embs + noise, token_embs], axis=0)
cache_b, first_b = make_cache_with_embeddings(combined_noisy)

# --- Test C: zero vectors ---
combined_zeros = mx.concatenate([mx.zeros((4, embed_dim)), token_embs], axis=0)
cache_c, first_c = make_cache_with_embeddings(combined_zeros)

# Warmup
for c, ft in [(cache_a, first_a), (cache_b, first_b), (cache_c, first_c)]:
    raw_decode(copy.deepcopy(c), ft, 8)

# Benchmark
for label, c, ft in [
    ("A: exact vocab embeddings", cache_a, first_a),
    ("B: vocab + 0.01 noise", cache_b, first_b),
    ("C: zero vectors", cache_c, first_c),
]:
    cc = copy.deepcopy(c)
    t0 = time.perf_counter()
    toks = raw_decode(cc, ft, 64)
    elapsed = time.perf_counter() - t0
    print(f"{label}: {elapsed:.3f}s ({64/elapsed:.1f} tok/s)")
```

## Results

| Soft prompt content | Decode tok/s |
|---|---:|
| Exact vocab embeddings | **93.1** |
| Vocab + 0.01 noise | **34.4** |
| Zero vectors | **34.5** |
| Random normalized | **34.3** |
| No soft prompt (baseline) | **93.2** |

## Analysis

The slowdown is in the **decode phase**, not prefill. Once the DeltaNet recurrent state has been computed from non-vocabulary embeddings, every subsequent `model(token, cache=cache)` call during autoregressive decoding runs ~2.7x slower.

The Gated DeltaNet layers (30 of 40 layers in Qwen3.5-35B-A3B) maintain a fixed-size recurrent state `(batch, heads, key_dim, val_dim)`. When this state is computed from embeddings that lie off the vocabulary manifold, it appears to trigger a slower Metal kernel path for all subsequent decode steps.

Pure attention layers (every 4th layer) are not affected — models without DeltaNet layers show zero overhead with `input_embeddings`.

## Expected behavior

Decode throughput should be independent of how the KV/recurrent cache was initially filled, since each decode step processes a single vocabulary token regardless.

## Environment

- **Hardware**: Mac Studio, Apple M4 Max, 36 GB
- **macOS**: 26.2 (25C56)
- **mlx**: 0.30.6
- **mlx-lm**: 0.30.7
- **Model**: `NexVeridian/Qwen3.5-35B-A3B-4bit`
- **Python**: 3.12

**Comment by angeloskath** [comment]
In all cases the inputs become float32 instead of the original type which causes the decoding to be more than 2x slower.

If you cast it back to the dtype of the original embeddings it should be fine.

**Comment by josaum** [comment]
thanks!
